File: dart/lib/src/tokenize.dart
import 'fixation.dart';
import 'options.dart';

/// A piece of the input: a [WordToken] or a [SeparatorToken].
///
/// Concatenating `token.text` over the list returned by [tokenize] gives the
/// input back unchanged. The class is sealed, so a `switch` over tokens is
/// exhaustive, which is what a Flutter `TextSpan` builder wants.
sealed class Token {
  const Token._(this.text);

  /// The exact slice of the input this token covers.
  final String text;
}

/// A word, with its fixation split.
final class WordToken extends Token {
  /// Creates a word token. `fixationText + restText` must equal `text`.
  const WordToken(
    super.text, {
    required this.fixation,
    required this.fixationText,
    required this.restText,
  }) : super._();

  /// A word that receives no emphasis: [fixation] `0`, all text in [restText].
  const WordToken.plain(String text)
      : this(text, fixation: 0, fixationText: '', restText: text);

  /// Number of emphasised leading grapheme clusters; `0` when the word gets
  /// no fixation (suppressed by a rule, or off the saccade beat).
  final int fixation;

  /// The emphasised prefix (empty when [fixation] is `0`).
  final String fixationText;

  /// The rest of the word (the whole word when [fixation] is `0`).
  final String restText;

  @override
  bool operator ==(Object other) =>
      other is WordToken &&
      other.text == text &&
      other.fixation == fixation &&
      other.fixationText == fixationText &&
      other.restText == restText;

  @override
  int get hashCode => Object.hash(text, fixation, fixationText, restText);

  @override
  String toString() =>
      'WordToken($text, fixation: $fixation, "$fixationText" + "$restText")';
}

/// Whitespace, punctuation, emoji or other non-word text between words.
final class SeparatorToken extends Token {
  /// Creates a separator token.
  const SeparatorToken(super.text) : super._();

  @override
  bool operator ==(Object other) =>
      other is SeparatorToken && other.text == text;

  @override
  int get hashCode => Object.hash(SeparatorToken, text);

  @override
  String toString() => 'SeparatorToken($text)';
}

/// Saccade bookkeeping shared across several [tokenizeWithState] calls, so
/// [toHtml] keeps counting words across markup (SPEC §4).
final class TokenizeState {
  /// Creates a counter starting at [wordIndex].
  TokenizeState([this.wordIndex = 0]);

  /// Index of the next word token; the first word of the input is `0`.
  int wordIndex;
}

/// Splits [text] into words and separators and computes each word's fixation
/// prefix (SPEC §3, §4).
///
/// ```dart
/// tokenize('Smooth reading');
/// // [WordToken(Smooth, fixation: 3, "Smo" + "oth"),
/// //  SeparatorToken( ),
/// //  WordToken(reading, fixation: 4, "read" + "ing")]
/// ```
List<Token> tokenize(
  String text, [
  SmoothOptions options = SmoothOptions.defaults,
]) =>
    tokenizeWithState(text, options, TokenizeState());

/// [tokenize] with an externally owned saccade counter.
List<Token> tokenizeWithState(
  String text,
  SmoothOptions options,
  TokenizeState state,
) {
  final tokens = <Token>[];
  for (final segment in options.segmenter.segmentWords(text, options.locale)) {
    if (segment.text.isEmpty) continue;
    if (segment.isWord) {
      tokens.add(_wordToken(segment.text, options, state));
    } else if (tokens.isNotEmpty && tokens.last is SeparatorToken) {
      // Consecutive separators merge into one token so every segmenter
      // produces the same token stream.
      tokens.last = SeparatorToken(tokens.last.text + segment.text);
    } else {
      tokens.add(SeparatorToken(segment.text));
    }
  }
  return tokens;
}

WordToken _wordToken(String word, SmoothOptions options, TokenizeState state) {
  // Every word token consumes a saccade index, including numbers and words
  // too short to be emphasised (SPEC §4).
  final onSaccade = state.wordIndex % options.saccade == 0;
  state.wordIndex += 1;
  if (!onSaccade) return WordToken.plain(word);

  final graphemes = options.segmenter.graphemes(word, options.locale);
  final n = graphemes.length;
  final requested = options.fixationLength == null
      ? fixationLength(word, n, options)
      : options.fixationLength!(word, n, options);
  final fixation = requested < 0 ? 0 : (requested > n ? n : requested);
  if (fixation == 0) return WordToken.plain(word);
  return WordToken(
    word,
    fixation: fixation,
    fixationText: graphemes.take(fixation).join(),
    restText: graphemes.skip(fixation).join(),
  );
}

File: rust/src/tokenize.rs
//! Tokenization: words with their fixation split, and separators (SPEC §3, §4).

use crate::fixation::fixation_length;
use crate::options::Options;
use crate::segment::Segmenter;

/// A piece of the input: a word or a separator.
///
/// Tokens borrow from the input. Concatenating [`Token::text`] over the
/// slice returned by [`tokenize`] gives the input back unchanged, and for a
/// word `fixation_text` + `rest_text` == `text`.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Token<'a> {
    /// A word, with its fixation split.
    Word {
        /// The exact slice of the input this word covers.
        text: &'a str,
        /// Number of emphasised leading grapheme clusters; `0` when the word
        /// gets no fixation (suppressed by a rule, or off the saccade beat).
        fixation: usize,
        /// The emphasised prefix (empty when `fixation` is `0`).
        fixation_text: &'a str,
        /// The rest of the word (the whole word when `fixation` is `0`).
        rest_text: &'a str,
    },
    /// Whitespace, punctuation, emoji or other non-word text between words.
    Separator {
        /// The exact slice of the input this separator covers.
        text: &'a str,
    },
}

impl<'a> Token<'a> {
    /// The exact slice of the input this token covers.
    #[must_use]
    pub const fn text(&self) -> &'a str {
        match self {
            Self::Word { text, .. } | Self::Separator { text } => text,
        }
    }

    /// `true` for [`Token::Word`], `false` for [`Token::Separator`].
    #[must_use]
    pub const fn is_word(&self) -> bool {
        matches!(self, Self::Word { .. })
    }

    /// A word that receives no emphasis: `fixation` `0`, all text in `rest_text`.
    const fn plain(text: &'a str) -> Self {
        Self::Word {
            text,
            fixation: 0,
            fixation_text: "",
            rest_text: text,
        }
    }
}

/// Saccade bookkeeping shared across several [`tokenize_with_state`] calls,
/// so [`to_html`](crate::to_html) keeps counting words across markup (SPEC §4).
#[derive(Debug, Clone, Copy, Default, PartialEq, Eq)]
pub struct TokenizeState {
    /// Index of the next word token; the first word of the input is `0`.
    pub word_index: usize,
}

/// Splits `text` into words and separators and computes each word's fixation
/// prefix (SPEC §3, §4).
///
/// ```
/// use smooth_reading::{Options, Token, tokenize};
///
/// let tokens = tokenize("Smooth reading", &Options::new());
/// assert_eq!(
///     tokens,
///     [
///         Token::Word { text: "Smooth", fixation: 3, fixation_text: "Smo", rest_text: "oth" },
///         Token::Separator { text: " " },
///         Token::Word { text: "reading", fixation: 4, fixation_text: "read", rest_text: "ing" },
///     ]
/// );
/// ```
#[must_use]
pub fn tokenize<'a>(text: &'a str, options: &Options) -> Vec<Token<'a>> {
    tokenize_with_state(text, options, &mut TokenizeState::default())
}

/// [`tokenize`] with an externally owned saccade counter.
///
/// Every word token advances `state.word_index`, so tokenizing one text in
/// several pieces keeps counting words across the pieces.
///
/// ```
/// use smooth_reading::{Options, Token, TokenizeState, tokenize_with_state};
///
/// let options = Options::new().saccade(2);
/// let mut state = TokenizeState::default();
/// let first = tokenize_with_state("one two", &options, &mut state);
/// let second = tokenize_with_state("three four", &options, &mut state);
/// assert_eq!(state.word_index, 4);
/// assert!(matches!(first[0], Token::Word { fixation: 2, .. })); // "one", index 0
/// assert!(matches!(first[2], Token::Word { fixation: 0, .. })); // "two", index 1, off the beat
/// assert!(matches!(second[0], Token::Word { fixation: 3, .. })); // "three", index 2, on the beat
/// assert!(matches!(second[2], Token::Word { fixation: 0, .. })); // "four", index 3, off the beat
/// ```
#[must_use]
pub fn tokenize_with_state<'a>(
    text: &'a str,
    options: &Options,
    state: &mut TokenizeState,
) -> Vec<Token<'a>> {
    let segmenter = options.segmenter_impl();
    let mut tokens: Vec<Token<'a>> = Vec::new();
    for segment in segmenter.segment_words(text, options.get_locale()) {
        if segment.text.is_empty() {
            continue;
        }
        if segment.is_word {
            tokens.push(word_token(segment.text, options, segmenter, state));
            continue;
        }
        // Consecutive separators merge into one token so every segmenter
        // produces the same token stream.
        if let Some(Token::Separator { text: last }) = tokens.last_mut() {
            if let Some(merged) = merge_adjacent(text, last, segment.text) {
                *last = merged;
                continue;
            }
        }
        tokens.push(Token::Separator { text: segment.text });
    }
    tokens
}

/// The slice of `text` covering `first` immediately followed by `second`, when
/// both are adjacent slices of `text` (which every well-behaved segmenter
/// returns); `None` otherwise.
fn merge_adjacent<'a>(text: &'a str, first: &str, second: &str) -> Option<&'a str> {
    let base = text.as_ptr().addr();
    let first_start = first.as_ptr().addr();
    let second_start = second.as_ptr().addr();
    if first_start < base
        || first_start + first.len() != second_start
        || second_start + second.len() > base + text.len()
    {
        return None;
    }
    let start = first_start - base;
    text.get(start..start + first.len() + second.len())
}

fn word_token<'a>(
    word: &'a str,
    options: &Options,
    segmenter: &dyn Segmenter,
    state: &mut TokenizeState,
) -> Token<'a> {
    // Every word token consumes a saccade index, including numbers and words
    // too short to be emphasised (SPEC §4).
    let on_saccade = state.word_index % options.get_saccade() == 0;
    state.word_index += 1;
    if !on_saccade {
        return Token::plain(word);
    }

    let graphemes = segmenter.graphemes(word, options.get_locale());
    let fixation = fixation_length(word, graphemes.len(), options).min(graphemes.len());
    if fixation == 0 {
        return Token::plain(word);
    }
    let split: usize = graphemes[..fixation].iter().map(|g| g.len()).sum();
    match (word.get(..split), word.get(split..)) {
        (Some(fixation_text), Some(rest_text)) => Token::Word {
            text: word,
            fixation,
            fixation_text,
            rest_text,
        },
        // Only reachable with a segmenter whose clusters are not slices of the
        // word; the word is then emitted unemphasised rather than mis-split.
        _ => Token::plain(word),
    }
}
